bot.go

```diff
@@ -557,21 +557,44 @@ func (bot *BotAPI) ListenForWebhook(pattern string) UpdatesChannel {
 	ch := make(chan Update, bot.Buffer)
 
 	http.HandleFunc(pattern, func(w http.ResponseWriter, r *http.Request) {
-		ch <- bot.HandleUpdate(w, r)
+		update, err := bot.HandleUpdate(r)
+		if err != nil {
+			errMsg, _ := json.Marshal(map[string]string{"error": err.Error()})
+			w.WriteHeader(http.StatusBadRequest)
+			w.Header().Set("Content-Type", "application/json")
+			_, _ = w.Write(errMsg)
+			return
+		}
+
+		ch <- *update
 	})
 
 	return ch
 }
 
 // HandleUpdate parses and returns update received via webhook
-func (bot *BotAPI) HandleUpdate(res http.ResponseWriter, req *http.Request) Update {
-  bytes, _ := ioutil.ReadAll(req.Body)
-  req.Body.Close()
+func (bot *BotAPI) HandleUpdate(r *http.Request) (*Update, error) {
+	if r.Method != http.MethodPost {
+		err := errors.New("wrong HTTP method required POST")
+		return nil, err
+	}
 
-  var update Update
-  json.Unmarshal(bytes, &update)
+	payload, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := r.Body.Close(); err != nil {
+		return nil, err
+	}
+
+	var update Update
+	err = json.Unmarshal(payload, &update)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
While we are here, how about we use `json.Decoder` instead of `json.Unmarshal`? This way we don't need to load the entire body into memory first. Well, loading it into memory is [essentially what Go does internally](https://github.com/golang/go/blob/85afa2eb190d5d1a06584803bde4b4ee9b0e79b0/src/encoding/json/stream.go#L49-L79), but I think it is better to let the language worry about the implementation of decoding a JSON stream. (Plus, it has some other niceties like failing early if a token is not allowed as it's streamed in, closing the body stream for us, etc.)

Could do something like:

```suggestion
	var update Update
	err := json.NewDecoder(r.Body).Decode(&update)
	if err != nil {
		return nil, err
	}
```